### PR TITLE
Pass referrer from the viewer

### PR DESCRIPTION
**Pull request by dvoytenko**

### CODE REVIEW

File: src/service/viewer-impl.js [hunk 1]
```diff
@@ -392,6 +392,41 @@ export class Viewer {
...
+   */
+  getReferrerUrl() {
+    // If not embedded in a IFrame always return the document's referrer.
+    if (!this.isEmbedded()) {
+      return this.win.document.referrer;
+    }
+
+    // Viewer override.
+    if (this.params_['referrer']) {
+      return this.params_['referrer'];
```
Comment (cramforce) [inline src/service/viewer-impl.js]: We should only trust this for trusted viewers. That is a concept we need to introduce.

Basically if the referrer/ancestorOrigins is google.com, we'd do this.

Comment (dvoytenko) [inline src/service/viewer-impl.js]: Done. PTAL.

File: src/service/viewer-impl.js [hunk 2]
```diff
@@ -74,6 +74,12 @@ export const VisibilityState = {
 };
 
 
+/** @export {!Array<string>} */
+export const TRUSTED_VIEWER_HOSTS = [
+  'google.com'
```
Comment (cramforce) [inline src/service/viewer-impl.js]: This is where it gets complicated due to TLDs.

I'm not sure there is a way to do this without a full list of TLDs :(

File: src/service/viewer-impl.js [hunk 3]
```diff
@@ -74,6 +74,12 @@ export const VisibilityState = {
 };
 
 
+/** @export {!Array<string>} */
+export const TRUSTED_VIEWER_HOSTS = [
```
Comment (cramforce) [inline src/service/viewer-impl.js]: Add docs what the rough process is to add a viewer here (file an issue and we discuss).

File: src/service/viewer-impl.js [hunk 4]
```diff
@@ -392,6 +398,80 @@ export class Viewer {
...
+   * Whether the viewer has been whitelisted for more sensitive operations
+   * such as customizing referrer.
+   * @return {boolean}
+   */
+  isTrustedViewer() {
+    if (this.win.document.referrer &&
+            this.isTrustedViewerOrigin_(this.win.document.referrer)) {
+      return true;
+    }
+    try {
```
Comment (cramforce) [inline src/service/viewer-impl.js]: this should never fail.

File: test/functional/test-viewer.js [hunk 5]
```diff
@@ -211,4 +211,118 @@ describe('Viewer', () => {
...
+      windowApi.location.ancestorOrigins = [referrer];
+      expect(viewer.isTrustedViewer()).to.equal(toBeTrusted, 'by ancestor');
+      windowApi.location.ancestorOrigins = null;
+    }
+
+    it('should flag viewer as not trusted without referrer or ancestor', () => {
+      test('', false);
+    });
+
+    it('should tust host as referrer with https', () => {
```
Comment (cramforce) [inline test/functional/test-viewer.js]: nit: trust

File: test/functional/test-viewer.js [hunk 6]
```diff
@@ -211,4 +211,118 @@ describe('Viewer', () => {
...
+      test('https://google.com', true);
```
Comment (cramforce) [inline test/functional/test-viewer.js]: Add `www.google.com`

Do we trust `news.google.com`?

File: test/functional/test-viewer.js [hunk 7]
```diff
@@ -211,4 +211,118 @@ describe('Viewer', () => {
...
+    });
+
+    it('should not trust host as referrer with http', () => {
+      test('http://google.com', false);
+    });
+
+    it('should not tust non-whitelisted host as referrer', () => {
+      test('https://acme.com', false);
```
Comment (cramforce) [inline test/functional/test-viewer.js]: Cases
google.evil.com
www.google.evil.com
withgoogle.com

File: src/service/viewer-impl.js [hunk 8]
```diff
@@ -392,6 +398,80 @@ export class Viewer {
...
+   * @param {string} urlString
+   * @return {boolean}
+   * @private
+   */
+  isTrustedViewerOrigin_(urlString) {
+    const url = parseUrl(urlString);
+    for (let i = 0; i < TRUSTED_VIEWER_HOSTS.length; i++) {
+      const th = TRUSTED_VIEWER_HOSTS[i];
+      if (url.protocol == 'https:' &&
+              url.host.indexOf(th) == url.host.length - th.length) {
```
Comment (cramforce) [inline src/service/viewer-impl.js]: Did you chose `host` over `hostname`?

File: src/service/viewer-impl.js (hunk 2 above)
Comment (dvoytenko) [inline src/service/viewer-impl.js]: Done via regexp. The security review is requested as well in b/26930262

File: src/service/viewer-impl.js (hunk 3 above)
Comment (dvoytenko) [inline src/service/viewer-impl.js]: Done

File: src/service/viewer-impl.js (hunk 4 above)
Comment (dvoytenko) [inline src/service/viewer-impl.js]: Done

File: test/functional/test-viewer.js (hunk 5 above)
Comment (dvoytenko) [inline test/functional/test-viewer.js]: Done

File: test/functional/test-viewer.js (hunk 6 above)
Comment (dvoytenko) [inline test/functional/test-viewer.js]: Done

File: test/functional/test-viewer.js (hunk 7 above)
Comment (dvoytenko) [inline test/functional/test-viewer.js]: Done

File: src/service/viewer-impl.js (hunk 8 above)
Comment (dvoytenko) [inline src/service/viewer-impl.js]: Done

**Comment by dvoytenko** [comment]
@cramforce PTAL

File: extensions/amp-dynamic-css-classes/0.1/test/test-runtime-classes.js [hunk 9]
```diff
@@ -57,6 +58,7 @@ describe('dynamic classes are inserted at runtime', () => {
       }
       if (referrer !== undefined) {
         overwrite(fixture.doc, 'referrer', referrer);
```
Comment (jridgewell) [inline extensions/amp-dynamic-css-classes/0.1/test/test-runtime-classes.js]: I think this line can be removed?

File: src/service/viewer-impl.js [hunk 10]
```diff
@@ -392,6 +409,84 @@ export class Viewer {
...
+  /**
+   * @param {string} urlString
+   * @return {boolean}
+   * @private
+   */
+  isTrustedViewerOrigin_(urlString) {
+    const url = parseUrl(urlString);
+    for (let i = 0; i < TRUSTED_VIEWER_HOSTS.length; i++) {
+      const th = TRUSTED_VIEWER_HOSTS[i];
+      if (url.protocol == 'https:' && th.test(url.hostname)) {
```
Comment (jridgewell) [inline src/service/viewer-impl.js]: We can short circuit without the for-loop for non-HTTPS urls.

File: src/service/viewer-impl.js [hunk 11]
```diff
@@ -392,6 +409,84 @@ export class Viewer {
...
+  }
+
+  /**
+   * @param {string} urlString
+   * @return {boolean}
+   * @private
+   */
+  isTrustedViewerOrigin_(urlString) {
+    const url = parseUrl(urlString);
+    for (let i = 0; i < TRUSTED_VIEWER_HOSTS.length; i++) {
```
Comment (jridgewell) [inline src/service/viewer-impl.js]: `Array#some` might be clearer here?

Comment (cramforce) [inline src/service/viewer-impl.js]: Not available I believe.

Comment (jridgewell) [inline src/service/viewer-impl.js]: IE9+? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some#Browser_compatibility

File: extensions/amp-dynamic-css-classes/0.1/test/test-runtime-classes.js (hunk 9 above)
Comment (dvoytenko) [inline extensions/amp-dynamic-css-classes/0.1/test/test-runtime-classes.js]: Correct. Removed.

File: src/service/viewer-impl.js (hunk 10 above)
Comment (dvoytenko) [inline src/service/viewer-impl.js]: Done.

File: build-system/tasks/presubmit-checks.js [hunk 12]
```diff
@@ -243,7 +243,15 @@ var forbiddenTerms = {
     whitelist: [
       'extensions/amp-access/0.1/access-expr-impl.js',
     ],
-  }
+  },
+
+  // Overridden APIs.
```
Comment (cramforce) [inline build-system/tasks/presubmit-checks.js]: lets also blacklist isTrusted

File: src/service/viewer-impl.js (hunk 11 above)
Comment (dvoytenko) [inline src/service/viewer-impl.js]: I think at the latest we decided not to use it, but I see that we are already using it elsewhere. Replaced with `some`.

File: test/functional/test-viewer.js [hunk 13]
```diff
@@ -211,4 +213,139 @@ describe('Viewer', () => {
...
+      test('https://news.google.com', true);
+      test('https://www.google.co.uk', true);
+      test('https://www.google.co.au', true);
+      test('https://news.google.co.uk', true);
+      test('https://news.google.co.au', true);
+    });
+
+    it('should NOT trust wrong or non-whitelisted domain variations', () => {
+      test('https://google.other.com', false);
+      test('https://google.net', false);
```
Comment (cramforce) [inline test/functional/test-viewer.js]: This fails? Does `google.de` pass?

File: build-system/tasks/presubmit-checks.js (hunk 12 above)
Comment (dvoytenko) [inline build-system/tasks/presubmit-checks.js]: Done

File: test/functional/test-viewer.js (hunk 7 above)
Comment (cramforce) [inline test/functional/test-viewer.js]: I don't believe you have all these cases yet.

File: test/functional/test-viewer.js (hunk 13 above)
Comment (dvoytenko) [inline test/functional/test-viewer.js]: `google.de` will currently fail. I guess I should let it pass... Will do.

Comment (dvoytenko) [inline test/functional/test-viewer.js]: Wait. I'm wrong. `google.de` should be fine. I added the test for it as well.

**Comment by dvoytenko** [comment]
@cramforce PTAL.

File: test/functional/test-viewer.js (hunk 13 above)
Comment (cramforce) [inline test/functional/test-viewer.js]: Please see request for more test cases I sent.

Would also be good to have more good cases.

Comment (dvoytenko) [inline test/functional/test-viewer.js]: Done: added more negative and positive tests as advised.

**Comment by cramforce** [comment]
LGTM

**Comment by dvoytenko** [comment]
I will merge the new cases with taboola here as well.

**Pull request merged**